sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala

```diff
@@ -96,6 +98,31 @@ class CollationSQLRegexpSuite
     }
   }
 
+  test("RegExpReplace throws the right exception when replace fails on a particular row") {
+    val tableName = "regexpReplaceException"
+    withTable(tableName) {
+      Seq("NO_CODEGEN", "CODEGEN_ONLY").foreach { codegenMode =>
+        withSQLConf("spark.sql.codegen.factoryMode" -> codegenMode) {
+          sql(s"CREATE TABLE IF NOT EXISTS $tableName(s STRING)")
+          sql(s"INSERT INTO $tableName VALUES('first last')")
+          val query = s"SELECT regexp_replace(s, '(?<first>[a-zA-Z]+) (?<last>[a-zA-Z]+)', " +
+            s"'$$3 $$1') FROM $tableName"
```

Comment:
I don't know. Is there any other databases support `regexp_replace(s, '(?<first>[a-zA-Z]+) (?<last>[a-zA-Z]+)', '$$3 $$1')` ?

Reply:
I guess the designer of `RegExpReplace` didn't realize the feature(In fact, a bug here). We should forbid it or consider a better way.

Reply:
Personally, I think check the rep expression if contains $1, $2 ... is a easier method.

Reply:
@beliefer The designer of `RegExpReplace` very well may have thought of this case and in my opinion it works exactly as intended. It's just that sometimes these patterns are out of index and we just throw the library error in this case. This PR aims to handle these exceptions better.